File: cloud/src/xml/probe.ts
import { XMLParser } from "fast-xml-parser";

export type ProbeDataItem = {
  id: string;
  name?: string;
  category: "SAMPLE" | "EVENT" | "CONDITION";
  type: string;
  subType?: string;
  units?: string;
  nativeUnits?: string;
  componentPath: string;
};

export type ProbeDevice = {
  uuid: string;
  name: string;
  model?: string;
  dataItems: ProbeDataItem[];
};

export type ProbeParseResult = {
  header: {
    instanceId: string;
    schemaVersion: string;
    creationTime: string;
  };
  devices: ProbeDevice[];
};

const parser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  parseAttributeValue: false,
  allowBooleanAttributes: true,
  isArray: (tagName) => {
    // tags that should always be arrays even with a single element
    return [
      "Device",
      "DataItem",
      "Components",
      "Axes",
      "Controller",
      "Path",
      "Linear",
      "Rotary",
      "Systems",
      "Auxiliaries",
      "Hydraulic",
      "Electric",
      "Pneumatic",
      "Coolant",
      "Lubrication",
    ].includes(tagName);
  },
});

export function parseProbe(xml: string): ProbeParseResult {
  const root = parser.parse(xml);
  const md = root.MTConnectDevices;
  const hdr = md.Header;

  const devices: ProbeDevice[] = [];
  const deviceArr = md.Devices?.Device ?? [];
  for (const d of deviceArr) {
    const dataItems: ProbeDataItem[] = [];
    collectDataItems(d, d["@_name"] ?? "", dataItems);
    devices.push({
      uuid: d["@_uuid"],
      name: d["@_name"],
      model: d["@_model"],
      dataItems,
    });
  }

  return {
    header: {
      instanceId: hdr["@_instanceId"],
      schemaVersion: hdr["@_schemaVersion"] ?? hdr["@_version"] ?? "",
      creationTime: hdr["@_creationTime"],
    },
    devices,
  };
}

function collectDataItems(
  node: Record<string, unknown>,
  path: string,
  out: ProbeDataItem[],
): void {
  const dis = (node as { DataItems?: { DataItem?: unknown[] } }).DataItems
    ?.DataItem;
  if (Array.isArray(dis)) {
    for (const di of dis as Array<Record<string, string>>) {
      out.push({
        id: di["@_id"],
        name: di["@_name"],
        category: di["@_category"] as "SAMPLE" | "EVENT" | "CONDITION",
        type: di["@_type"],
        subType: di["@_subType"],
        units: di["@_units"],
        nativeUnits: di["@_nativeUnits"],
        componentPath: path,
      });
    }
  }
  // recurse into any nested component tags
  // with isArray:["Components"], node.Components is an array of objects,
  // each object keyed by component-tag => array of child components.
  const components = (node as { Components?: unknown }).Components;
  const compArr = Array.isArray(components)
    ? components
    : components
      ? [components]
      : [];
  for (const compObj of compArr) {
    for (const [tag, children] of Object.entries(
      compObj as Record<string, unknown>,
    )) {
      // skip attribute-ish keys (shouldn't be any under Components, but be safe)
      if (tag.startsWith("@_")) continue;
      const childArr = Array.isArray(children) ? children : [children];
      for (const child of childArr as Array<Record<string, unknown>>) {
        if (!child || typeof child !== "object") continue;
        const name = (child as Record<string, string>)["@_name"] ?? tag;
        collectDataItems(child, `${path}/${name}`, out);
      }
    }
  }
}

File: edge/forwarder/src/forwarder/probe.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from lxml import etree

NS = "{urn:mtconnect.org:MTConnectDevices:2.7}"


@dataclass(frozen=True)
class DataItem:
    id: str
    name: str | None
    category: Literal["SAMPLE", "EVENT", "CONDITION"]
    type: str
    sub_type: str | None
    units: str | None
    native_units: str | None
    component_path: str


@dataclass(frozen=True)
class Device:
    uuid: str
    name: str
    model: str | None
    data_items: list[DataItem] = field(default_factory=list)


@dataclass(frozen=True)
class ProbeResult:
    instance_id: str
    schema_version: str
    creation_time: str
    devices: list[Device]


def _local(tag: str) -> str:
    return tag.split("}", 1)[-1] if "}" in tag else tag


def _collect_data_items(node: etree._Element, path: str, out: list[DataItem]) -> None:
    for child in node:
        local = _local(child.tag)
        if local == "DataItems":
            for di in child:
                if _local(di.tag) != "DataItem":
                    continue
                cat = di.get("category")
                if cat not in {"SAMPLE", "EVENT", "CONDITION"}:
                    continue
                out.append(
                    DataItem(
                        id=di.get("id", ""),
                        name=di.get("name"),
                        category=cat,  # type: ignore[arg-type]
                        type=di.get("type", ""),
                        sub_type=di.get("subType"),
                        units=di.get("units"),
                        native_units=di.get("nativeUnits"),
                        component_path=path,
                    )
                )
        elif local in {"Components"}:
            for comp in child:
                comp_name = comp.get("name") or _local(comp.tag)
                _collect_data_items(comp, f"{path}/{comp_name}", out)


def parse_probe(xml: str) -> ProbeResult:
    root = etree.fromstring(xml.encode("utf-8"))
    header = next(child for child in root if _local(child.tag) == "Header")
    devices_node = next(child for child in root if _local(child.tag) == "Devices")

    devices: list[Device] = []
    for d in devices_node:
        if _local(d.tag) != "Device":
            continue
        data_items: list[DataItem] = []
        _collect_data_items(d, d.get("name", ""), data_items)
        devices.append(
            Device(
                uuid=d.get("uuid", ""),
                name=d.get("name", ""),
                model=d.get("model"),
                data_items=data_items,
            )
        )

    return ProbeResult(
        instance_id=header.get("instanceId", ""),
        schema_version=header.get("schemaVersion", header.get("version", "")),
        creation_time=header.get("creationTime", ""),
        devices=devices,
    )
